File: src/examples/javascript/factorial.js
/**
 * Factorial - JavaScript
 *
 * Calculate factorial using recursion.
 * n! = n × (n-1) × (n-2) × ... × 1
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n) - call stack
 */

export default {
  id: 'factorial-js',
  language: 'javascript',
  title: 'Factorial (Recursive)',
  description: 'Calculate factorial using recursion',
  difficulty: 'beginner',
  category: 'recursion',
  tags: ['algorithm', 'recursion', 'mathematics'],
  code: `function factorial(n) {
  // Base case
  if (n === 0 || n === 1) {
    return 1;
  }
  
  // Recursive case
  return n * factorial(n - 1);
}

// Test the function
console.log("Factorial of 5:", factorial(5));
console.log("Factorial of 0:", factorial(0));
console.log("Factorial of 7:", factorial(7));

// Calculate factorials 1-10
for (let i = 1; i <= 10; i++) {
  console.log(\`\${i}! = \${factorial(i)}\`);
}`
};

File: src/examples/cpp/factorial.cpp
export default {
  id: 'factorial-cpp',
  language: 'cpp',
  title: 'Recursive Factorial (C++)',
  description: 'Compute factorial using recursion in C++',
  difficulty: 'beginner',
  category: 'recursion',
  tags: ['algorithm', 'recursion', 'mathematics'],
  code: `#include <iostream>
using namespace std;

long long factorial(int n) {
  if (n <= 1) return 1;
  return n * factorial(n - 1);
}

int main() {
  cout << "Factorials 1 through 10:" << endl;
  for (int i = 1; i <= 10; i++) {
    cout << i << "! = " << factorial(i) << endl;
  }
  return 0;
}`
};
